lib/ret/api/rooms.ex

```diff
@@ -0,0 +1,183 @@
+    if can?(credentials, create_room(nil)) do
+      do_create_room(credentials, params)
+    else
+      {:error, :invalid_credentials}
+    end
+  end
+
+  def authed_update_room(hub_sid, %Credentials{} = credentials, params) do
+    hub = Hub |> Repo.get_by(hub_sid: hub_sid) |> Repo.preload([:hub_role_memberships, :hub_bindings])
+
+    if is_nil(hub) do
+      {:error, "Cannot find room with id: " <> hub_sid}
+    else
+      if can?(credentials, update_room(hub)) do
+        do_update_room(hub, credentials, params)
+      else
+        {:error, :invalid_credentials}
+      end
+    end
+  end
+
+  defp do_update_room(hub, %Credentials{subject_type: :app}, params) do
+    hub
+    |> Hub.add_attrs_to_changeset(params)
+    |> Hub.maybe_add_member_permissions(hub, params)
+    |> Hub.maybe_add_new_scene_to_changeset(params)
+    |> try_do_update_room(:reticulum_app_token)
+  end
+
+  defp do_update_room(hub, %Credentials{subject_type: :account, account: account}, params) do
+    hub
+    |> Hub.add_attrs_to_changeset(params)
+    |> Hub.maybe_add_member_permissions(hub, params)
+    |> Hub.maybe_add_promotion(account, hub, params)
+    |> Hub.maybe_add_new_scene_to_changeset(params)
+    |> try_do_update_room(account)
+  end
+
+  defp try_do_update_room({:error, reason}, _) do
+    {:error, reason}
+  end
+
+  defp try_do_update_room(changeset, :reticulum_app_token) do
+    case changeset |> Repo.update() do
+      {:error, changeset} ->
+        {:error, changeset}
+
+      {:ok, hub} ->
+        hub = Repo.preload(hub, Hub.hub_preloads())
+
+        case broadcast_hub_refresh(hub, :reticulum_app_token) do
+          {:error, reason} -> {:error, reason}
+          :ok -> {:ok, hub}
+        end
+    end
+  end
+
+  defp try_do_update_room(changeset, account) do
+    case changeset |> Repo.update() do
+      {:error, changeset} ->
+        {:error, changeset}
+
+      {:ok, hub} ->
+        hub = Repo.preload(hub, Hub.hub_preloads())
+
+        case broadcast_hub_refresh(hub, account) do
+          {:error, reason} -> {:error, reason}
+          :ok -> {:ok, hub}
+        end
+    end
+  end
+
+  defp broadcast_hub_refresh(hub, :reticulum_app_token) do
+    payload =
+      HubView.render("show.json", %{
+        hub: hub,
+        embeddable: true
+      })
+      |> Map.put(:stale_fields, [
+        # TODO: Only include fields that have changed in stale_fields
+        "name",
+        "description",
+        "member_permissions",
+        "room_size",
+        "allow_promotion",
+        "scene"
+      ])
+
+    RetWeb.Endpoint.broadcast("hub:" <> hub.hub_sid, "hub_refresh", payload)
+  end
+
+  defp broadcast_hub_refresh(hub, %Account{} = account) do
+    payload =
+      HubView.render("show.json", %{
+        hub: hub,
+        embeddable: account |> can?(embed_hub(hub))
+      })
+      |> Map.put(:stale_fields, [
+        # TODO: Only include fields that have changed in stale_fields
```

Comment:
Yep. 

In hub_channel we do "manually":  https://github.com/mozilla/reticulum/blob/ec9cd680fa3b995a235d88d6de6cec950b318659/lib/ret_web/channels/hub_channel.ex#L414-L420 

I decided instead to inspect the `changeset` itself to figure out what changed:
https://github.com/mozilla/reticulum/blob/ec9cd680fa3b995a235d88d6de6cec950b318659/lib/ret/api/rooms.ex#L95

However this means that we are no longer grouping changes to `scene`, `scene_listing` and `default_environment_gltf_bundle_url` all under "scene" like we do here:
https://github.com/mozilla/reticulum/blob/ec9cd680fa3b995a235d88d6de6cec950b318659/lib/ret_web/channels/hub_channel.ex#L482

So that we'd have to add these lines to the client:
https://github.com/mozilla/hubs/pull/3391/files

Is this preferred, or should I be emitting a change that says the "scene" changed for any of these changes?

Reply:
As long as `changeset.changes` only include fields that actually changed (i.e. whose values are different from the hubs' existing values), this approach looks good.

Reply:
I believe this is the case
> "Changed attributes will only be added if the change does not have the same value as the field in the data."

```elixir
iex> changeset = change(%Post{author: "bar"}, title: "title")
iex> changeset.changes
%{title: "title"}

iex> changeset = change(%Post{title: "title"}, title: "title")
iex> changeset.changes
%{}
```

https://hexdocs.pm/ecto/Ecto.Changeset.html#change/2-examples